### ISSUE TITLE
Failed to gain exclusive lock to the Cloud Firestore client's offline persistence.

### ISSUE THREAD

**Issue by hakobast** [firebase/firebase-android-sdk#5001]
### Environment

  * Firebase Component: 
```
'firebase_analytics'                      : '21.2.2'
'firebase_crashlytics_gradle'             : '2.9.0'
'firebase_crashlytics'                    : '18.3.7'
'firebase_messageing'                     : '23.1.2'
'firebase_auth'                           : '22.0.0'
'firebase_storage'                        : '20.2.0'
'firestore_ktx'                           : '24.6.0'
'firebase_config_ktx'                     : '21.4.0'
'firebase_perf_ktx'                       : '20.3.2'
'functions_ktx'                           : '20.3.0'
'firebase_perf_gradle'                    : '1.4.2'
'firebase_core'                           : '21.1.1'
```

### Problem

We are constantly receiving spikes of crashes related to the Firestore. Seems like something related to the sqlite db lock. Messages say that could be multiprocess issue when `Application` class gets initialized by multiple processes but we do not have it configured explicitly. I don't know any other ways that it could happen.  On our devices we are not able to reproduce the issue. 

<details>

`Caused by java.lang.RuntimeException: java.lang.RuntimeException: Failed to gain exclusive lock to the Cloud Firestore client's offline persistence. This generally means you are using Cloud Firestore from multiple processes in your app. Keep in mind that multi-process Android apps execute the code in your Application class in all processes, so you may need to avoid initializing Cloud Firestore in your Application class. If you are intentionally using Cloud Firestore from multiple processes, you can only enable offline persistence (that is, call setPersistenceEnabled(true)) in one of them.
       at com.google.firebase.firestore.util.AsyncQueue$SynchronizedShutdownAwareExecutor.lambda$executeAndReportResult$1(AsyncQueue.java:30)
       at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:457)
       at java.util.concurrent.FutureTask.run(FutureTask.java:266)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
       at com.google.firebase.firestore.util.AsyncQueue$SynchronizedShutdownAwareExecutor$DelayedStartFactory.run(AsyncQueue.java:15)
       at java.lang.Thread.run(Thread.java:764)`

`Caused by android.database.sqlite.SQLiteDatabaseLockedException: database is locked (code 5): , while compiling: PRAGMA journal_mode
       at android.database.sqlite.SQLiteConnection.nativePrepareStatement(SQLiteConnection.java)
       at android.database.sqlite.SQLiteConnection.acquirePreparedStatement(SQLiteConnection.java:903)
       at android.database.sqlite.SQLiteConnection.executeForString(SQLiteConnection.java:648)
       at android.database.sqlite.SQLiteConnection.setJournalMode(SQLiteConnection.java:334)
       at android.database.sqlite.SQLiteConnection.setWalModeFromConfiguration(SQLiteConnection.java:306)
       at android.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:216)
       at android.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:194)
       at android.database.sqlite.SQLiteConnectionPool.openConnectionLocked(SQLiteConnectionPool.java:493)
       at android.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:200)
       at android.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:192)
       at android.database.sqlite.SQLiteDatabase.openInner(SQLiteDatabase.java:869)
       at android.database.sqlite.SQLiteDatabase.open(SQLiteDatabase.java:854)
       at android.database.sqlite.SQLiteDatabase.openDatabase(SQLiteDatabase.java:729)
       at android.database.sqlite.SQLiteDatabase.openDatabase(SQLiteDatabase.java:719)
       at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:295)
       at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:238)
       at com.google.firebase.firestore.local.SQLitePersistence.u1(SQLitePersistence.java:17)
       at com.google.firebase.firestore.core.ComponentProvider.initialize(ComponentProvider.java:68)
       at com.google.firebase.firestore.core.FirestoreClient.initialize(FirestoreClient.java:68)
       at com.google.firebase.firestore.core.FirestoreClient.lambda$new$0(FirestoreClient.java:48)
       at com.google.firebase.firestore.util.AsyncQueue.lambda$enqueue$2(AsyncQueue.java:2)
       at com.google.firebase.firestore.util.AsyncQueue$SynchronizedShutdownAwareExecutor.lambda$executeAndReportResult$1(AsyncQueue.java:16)
       at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:457)
       at java.util.concurrent.FutureTask.run(FutureTask.java:266)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
       at com.google.firebase.firestore.util.AsyncQueue$SynchronizedShutdownAwareExecutor$DelayedStartFactory.run(AsyncQueue.java:15)
       at java.lang.Thread.run(Thread.java:764)`

<summary> Stacktrace </summary>

</details>

<details>
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:tools="http://schemas.android.com/tools"
    package="***">

    <uses-permission
        android:name="android.permission.WRITE_EXTERNAL_STORAGE"
        android:maxSdkVersion="28" />
    <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
    <uses-permission android:name="android.permission.INTERNET" />

    <application
        android:name=".App"
        android:allowBackup="false"
        android:dataExtractionRules="@xml/data_extraction_rules"
        android:icon="@mipmap/ic_launcher"
        android:label="@string/app_name"
        android:localeConfig="@xml/locales_config"
        android:roundIcon="@mipmap/ic_launcher_round"
        android:supportsRtl="true"
        android:theme="@style/AppTheme">

        <activity
            android:name="***"
            android:exported="true"
            android:launchMode="singleTask"
            android:theme="@style/AppTheme.Launcher">

        </activity>
        <activity
            android:name="***"
            android:exported="true"
            android:launchMode="singleTask"
            android:theme="@style/AppTheme.Launcher">

        </activity>
        <activity
            android:name="***"
            android:exported="true"
            android:launchMode="singleTask"
            android:theme="@style/AppTheme.Launcher">

        </activity>

        <activity
            android:name=".ui.screen.splash.SplashActivity"
            android:exported="true"
            android:screenOrientation="userPortrait"
            android:theme="@style/AppTheme.Launcher">
            <intent-filter>
                <action android:name="android.intent.action.MAIN" />

                <category android:name="android.intent.category.LAUNCHER" />
            </intent-filter>
        </activity>
        <activity
            android:name=".ui.screen.main.MainActivity"
            android:screenOrientation="userPortrait"
            android:windowSoftInputMode="adjustPan" />

    </application>

</manifest>

<summary>Android manifest</summary
</details>

![Screenshot 2023-05-16 at 00 58 38](https://github.com/firebase/firebase-android-sdk/assets/2909194/59fe2d55-5b47-4d28-add3-1a753402d44a)

**Comment by cherylEnkidu** [comment]
Hi, 
Can you share with us the manifest for your Android App? There is a chance you use multi processes and you are not aware of that. In the meantime, these are some tickets experienced similar issue https://github.com/firebase/firebase-android-sdk/issues/3029  https://github.com/firebase/firebase-android-sdk/issues/953 you might get a hint from these tickets.

**Comment by hakobast** [comment]
> Hi,
> Can you share with us the manifest for your Android App? There is a chance you use multi processes and you are not aware of that. In the meantime, these are some tickets experienced similar issue #3029 #953 you might get a hint from these tickets.

Sure, Added to description.

**Comment by cherylEnkidu** [comment]
Hi @hakobast ,

I didn't find anything special in your configuration that would possibly lead to this problem. And without reproduced app we cannot further investigate on our side. Some suggestions that might be helpful is: 

- Search for `Thread` keyword , check if there is any use inside the codebase that might related.
- If it is happened on a specific device, run the app on a real device to reproduce it. 
- See if it is possible user install same app twice on their devices.

**Comment by hakobast** [comment]
> Hi @hakobast ,
> 
> 
> 
> I didn't find anything special in your configuration that would possibly lead to this problem. And without reproduced app we cannot further investigate on our side. Some suggestions that might be helpful is: 
> 
> 
> 
> - Search for `Thread` keyword , check if there is any use inside the codebase that might related.
> 
> - If it is happened on a specific device, run the app on a real device to reproduce it. 
> 
> - See if it is possible user install same app twice on their devices.

Thanks I will check. About installing twice in a same device:  how is it possible? Is there any specific feature that phone needs to have or way to check it via code?

**Comment by timobaehr** [comment]
Some of my users are facing the same issue.
> Internal error in Cloud Firestore (24.6.0).

24.6.1 is not fixing the error.

Please check especially 32 bit C++ commands. PRAGMA looks like a 32 bit issue at C++.

Affected Android versions:
* 6, 2%
* 8.0.0, 55%
* 8.1.0, 17%,
* 9, 16%
* 10, 5%
* 11, 2%
* 12, 3%

Affected Devices (all with 32 bit processor?):
* SAMSUNG Galaxy Tab A (2016), 48%
* SAMSUNG Galaxy A5 (2017), 2%
* SAMSUNG Galaxy J6/S7, S5 Mini, Tab S 10.5, A10 (each one 1%)
* Huawei Mate 10 Lite, 8%
* Honor 9 Lite, 7%
* Huawei P20 Lite, 5%
* Huawei P10 Lite, 3%
* Huawei MediaPad T5, 2%
* Huawei Y5 2019, 1%
* Lenovo Tab E7 , 5%
* Lenove P1060X, <1%
* Droi - FIH Foxconn IPlay_20, 2%
* Others, 10%

Affected users: 40, 304 reports

Stacktrace sent via Crashlytics:
```
Fatal Exception: java.lang.RuntimeException: Internal error in Cloud Firestore (24.6.0).
       at com.google.firebase.firestore.util.AsyncQueue.lambda$panic$3(AsyncQueue.java:19)
       at android.os.Handler.handleCallback(Handler.java:789)
       at android.os.Handler.dispatchMessage(Handler.java:98)
       at android.os.Looper.loop(Looper.java:164)
       at android.app.ActivityThread.main(ActivityThread.java:6944)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:327)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1374)

Caused by java.lang.RuntimeException: java.lang.RuntimeException: Failed to gain exclusive lock to the Cloud Firestore client's offline persistence. This generally means you are using Cloud Firestore from multiple processes in your app. Keep in mind that multi-process Android apps execute the code in your Application class in all processes, so you may need to avoid initializing Cloud Firestore in your Application class. If you are intentionally using Cloud Firestore from multiple processes, you can only enable offline persistence (that is, call setPersistenceEnabled(true)) in one of them.
       at com.google.firebase.firestore.util.AsyncQueue$SynchronizedShutdownAwareExecutor.lambda$executeAndReportResult$1(AsyncQueue.java:14)
       at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:457)
       at java.util.concurrent.FutureTask.run(FutureTask.java:266)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
       at com.google.firebase.firestore.util.AsyncQueue$SynchronizedShutdownAwareExecutor$DelayedStartFactory.run(AsyncQueue.java:15)
       at java.lang.Thread.run(Thread.java:764)

Caused by java.lang.RuntimeException: Failed to gain exclusive lock to the Cloud Firestore client's offline persistence. This generally means you are using Cloud Firestore from multiple processes in your app. Keep in mind that multi-process Android apps execute the code in your Application class in all processes, so you may need to avoid initializing Cloud Firestore in your Application class. If you are intentionally using Cloud Firestore from multiple processes, you can only enable offline persistence (that is, call setPersistenceEnabled(true)) in one of them.
       at com.google.firebase.firestore.local.SQLitePersistence.start(SQLitePersistence.java:44)
       at com.google.firebase.firestore.core.ComponentProvider.initialize(ComponentProvider.java:6)
       at com.google.firebase.firestore.core.FirestoreClient.initialize(FirestoreClient.java:67)
       at com.google.firebase.firestore.core.FirestoreClient.lambda$new$0(FirestoreClient.java:10)
       at com.google.firebase.firestore.util.AsyncQueue.lambda$enqueue$2(AsyncQueue.java)
       at com.google.firebase.firestore.util.AsyncQueue$SynchronizedShutdownAwareExecutor.lambda$executeAndReportResult$1(AsyncQueue.java)
       at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:457)
       at java.util.concurrent.FutureTask.run(FutureTask.java:266)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
       at com.google.firebase.firestore.util.AsyncQueue$SynchronizedShutdownAwareExecutor$DelayedStartFactory.run(AsyncQueue.java:15)
       at java.lang.Thread.run(Thread.java:764)

Caused by android.database.sqlite.SQLiteDatabaseLockedException: database is locked (code 5): , while compiling: PRAGMA journal_mode
#################################################################
Error Code : 5 (SQLITE_BUSY)
Caused By : The database file is locked.
	(database is locked (code 5): , while compiling: PRAGMA journal_mode)
#################################################################
       at android.database.sqlite.SQLiteConnection.nativePrepareStatement(SQLiteConnection.java)
       at android.database.sqlite.SQLiteConnection.acquirePreparedStatement(SQLiteConnection.java:1096)
       at android.database.sqlite.SQLiteConnection.executeForString(SQLiteConnection.java:800)
       at android.database.sqlite.SQLiteConnection.setJournalMode(SQLiteConnection.java:449)
       at android.database.sqlite.SQLiteConnection.setWalModeFromConfiguration(SQLiteConnection.java:423)
       at android.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:272)
       at android.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:205)
       at android.database.sqlite.SQLiteConnectionPool.openConnectionLocked(SQLiteConnectionPool.java:649)
       at android.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:255)
       at android.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:222)
       at android.database.sqlite.SQLiteDatabase.openInner(SQLiteDatabase.java:1191)
       at android.database.sqlite.SQLiteDatabase.open(SQLiteDatabase.java:1146)
       at android.database.sqlite.SQLiteDatabase.openDatabase(SQLiteDatabase.java:861)
       at android.app.ContextImpl.openOrCreateDatabase(ContextImpl.java:729)
       at android.content.ContextWrapper.openOrCreateDatabase(ContextWrapper.java:310)
       at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:254)
       at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:194)
       at com.google.firebase.firestore.local.SQLitePersistence.start(SQLitePersistence.java:16)
       at com.google.firebase.firestore.core.ComponentProvider.initialize(ComponentProvider.java:6)
       at com.google.firebase.firestore.core.FirestoreClient.initialize(FirestoreClient.java:67)
       at com.google.firebase.firestore.core.FirestoreClient.lambda$new$0(FirestoreClient.java:10)
       at com.google.firebase.firestore.util.AsyncQueue.lambda$enqueue$2(AsyncQueue.java)
       at com.google.firebase.firestore.util.AsyncQueue$SynchronizedShutdownAwareExecutor.lambda$executeAndReportResult$1(AsyncQueue.java)
       at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:457)
       at java.util.concurrent.FutureTask.run(FutureTask.java:266)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
       at com.google.firebase.firestore.util.AsyncQueue$SynchronizedShutdownAwareExecutor$DelayedStartFactory.run(AsyncQueue.java:15)
       at java.lang.Thread.run(Thread.java:764)
```

**Comment by timobaehr** [comment]
> It probably means you're creating multiple DbUpdateHelper instances. You almost always only want to make one, ever.

> Found how to solve it. The problem for me was that I created multiple instances of SQLiteOpenHelper for the same database file from multiple threads. getWritableDatabase() contains synchronized block inside to prevent parallel access to internal open() method from multiple threads but because as I said there were multiple instances of SQLiteOpenHelper this lock didn't work. I created singleton from SQLiteOpenHelper. In order to release memory, I used WeakReference for "instance" field. Also, you probably want to add enableWriteAheadLogging() in order to access db methods in parallel.

Compare https://github.com/greenrobot/greenDAO/issues/191

**Comment by timobaehr** [comment]
Maybe there is a connection to [this issue](https://github.com/firebase/firebase-android-sdk/issues/4034).

I found a workaround. You have to disable the persistence on devices with 32 bit processor.
```
FirebaseFirestore.instance.settings = FirebaseFirestore.instance.settings.copyWith(
  persistenceEnabled: false,
);
```

I'm using this code now, so I'm not checking if the device architecture so far. Crash is no longer happening then.
```
    final androidInfoInfo = await DeviceInfoPlugin().androidInfo;
    const ANDROID_12_LEVEL = 32;
    final isAndroid12OrLess = androidInfoInfo.version.sdkInt <= ANDROID_12_LEVEL;
    if (isAndroid12OrLess) {
      FirebaseFirestore.instance.settings = FirebaseFirestore.instance.settings.copyWith(
        persistenceEnabled: false,
      );
    }
```

**Comment by hakobast** [comment]
@timobaehr thanks for the response, what is the issue with 32 bit processors?

**Comment by timobaehr** [comment]
@hakobast It is only a guess or an assumption:
`The database file is locked. (database is locked (code 5): , while compiling: PRAGMA journal_mode)`

**Comment by hakobast** [comment]
@timobaehr Recently we have [this](https://www.gsmarena.com/oppo_a7-9394.php) device experiencing the issue and it seems it has processor with 64 bit architecture.

**Comment by cherylEnkidu** [comment]
Hi @timobaehr , could you please share with me the reference you reach this conclusion? 
```
Please check especially 32 bit C++ commands. PRAGMA looks like a 32 bit issue at C++.
```
Also, it would be helpful if @hakobast and @timobaehr can help us find out starting which SDK version it leads to this problem. 

Thank you!

**Comment by timobaehr** [comment]
I posted a list of affected devices and looks like they all have a 32 bit processor. That's why I reach this conclusion.

Affected Android FireStore versions: `24.6.0`, `24.6.1`

Flutter plugin: `cloud_firestore 4.8.2`

I failed to dowgrade FireStore Android version, I tried with
```gradle
com.google.firebase:firebase-firestore-ktx:24.4.5
```
but it still used 24.6.0. But I would expect that 24.5.*/24.4.* was okay.

**Comment by hakobast** [comment]
@cherylEnkidu I cannot see any version specifically that does not have the issue

**Comment by rakshitsoni02** [comment]
hey @timobaehr @hakobast is there any fixed version released?

**Comment by timobaehr** [comment]
I didn't tried, it is not good to test this with users on production... I disabled the offline functionality.

**Comment by lehcar09** [comment]
Hi folks! sorry for the silence here. This issue does not relevant anymore due to deprecation of `setPersistenceEnabled` API since version [24.6.0](https://firebase.google.com/support/release-notes/android#firestore_v24-6-0).

Please check this [documentation](https://firebase.google.com/docs/firestore/manage-data/enable-offline) to configure Firestore offline features.

I'll be closing this issue not. Feel free to file new issue if you encounter issues. Thanks!
